pwa.php

```diff
@@ -253,6 +253,28 @@ function _pwa_check_disabled_navigation_preload() {
 /** Hooks to add for when accessing admin. */
 require_once PWA_PLUGIN_DIR . '/wp-admin/admin.php';
 
+/**
+ * Plugin activation hook.
+ */
+function _pwa_activate_plugin() {
+	pwa_add_rewrite_rules();
+	flush_rewrite_rules( false );
+}
+
+register_activation_hook( PWA_PLUGIN_FILE, '_pwa_activate_plugin' );
```

Comment:
The flushing is only happening upon activation. What about on network-activted multisite?

Reply:
I've just inquired the same on wp-sitemaps: https://github.com/GoogleChromeLabs/wp-sitemaps/issues/205

Reply:
Haven‘t really tested Multisite, will take a look.

Reply:
https://jeremyfelt.com/2015/07/17/flushing-rewrite-rules-in-wordpress-multisite-for-fun-and-profit/ 🤯

Reply:
cc @felixarntz for Multisite expertise.

Reply:
Iterating through all sites like in the above post works for small enough networks, but quickly runs into scaling issues. I think the only scalable way for a plugin to deal with setup like this in multisite is to not use an activation hook, but do it conditionally on something like `admin_init`:
* A common approach is to add an option for the plugin's version (e.g. we have `googlesitekit_db_version` in Site Kit).
* If version in DB does not exist or is lower than the current plugin version, run according installation / migration logic.

For installing rewrite rules like here a simple boolean flag would work, but I'd recommend something version-based because it's always likely to need more logic in the future which then will become migrations from older versions.

Reply:
Currently in the XML Sitemaps plugin it is [flushing](https://github.com/GoogleChromeLabs/wp-sitemaps/blob/584efba4eb63e918bc691c92bc573bba8945f07b/inc/class-wp-sitemaps.php#L166-L170) when the `WP_SITEMAPS_REWRITE_VERSION` changes [at `wp_loaded`](https://github.com/GoogleChromeLabs/wp-sitemaps/blob/584efba4eb63e918bc691c92bc573bba8945f07b/inc/class-wp-sitemaps.php#L68). That should rather have run at `admin_init`, no? In any case, that won't be needed if `flush_rewrite_rules9) ` is added to `upgrade_550()` as you noted in https://github.com/GoogleChromeLabs/wp-sitemaps/issues/205#issuecomment-640115945. The same could be done in the future if and when PWA is merged into core.

In the meantime, however, we should amend this PR with the `admin_init` hook that conditionally flushes rewrite rules if need be. As opposed to storing the version in an option like the XML Sitemap plugin is doing, how about something like this:

```php
add_action( 'admin_init', function() {
	/* @var WP_Rewrite $wp_rewrite */
	global $wp_rewrite;
	if ( ! isset( $wp_rewrite->extra_rules_top['^wp-service-worker\.js$'] ) ) {
		flush_rewrite_rules( false );
	}
} );
```

Reply:
I don't think this will work since `$wp_rewrite->extra_rules_top['^wp-service-worker\.js$']` will always be set by `pwa_add_rewrite_rules` on `init`.

Reply:
Good point. In that case, perhaps `pwa_add_rewrite_rules()` should check to see if it actually added anything to the array, and if so, set a flag that can be accessed somewhere?

Reply:
:wave: y'all!

FWIW, since that post was written I've had success with the method posted by Thorsten on this ticket: https://core.trac.wordpress.org/ticket/18450

I've used it on `init` when the rewrite rules were very important, but it could be used on `admin_init` as well.

Reply:
Thanks for the input! Yes, I'd say this qualifies as being very important since if the PWA plugin is network-activated, all the network sites would try to install the service worker. But if the rewrite rule is not present, then they'd fail to install.

Nevertheless, we're currently blocked on using rewrite rule, at least using `.js` file extension since various hosts will short-circuit such requests from going to WordPress to instead try to go straight to the filesystem: https://github.com/GoogleChromeLabs/pwa-wp/issues/287#issuecomment-642946205

Reply:
Now that we have `/wp.serviceworker` as the route, the only thing left here is making sure that the route will be detected across a network before rewrite rules have been flushed. 

Here's an idea: en lieu of waiting for someone to log-in to each site to cause rewrite rules to be flushed, we can implement a fallback route of sorts. 

I think I've got a good solution for that in 043beb8c29671cf3a32d40cf9d7bc5ce606a4884.